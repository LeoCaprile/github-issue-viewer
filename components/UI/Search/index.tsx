import Button from '@components/UI/Button';
import useDefaultValue from 'hooks';
import { useRouter } from 'next/router';
import { FormEvent, useRef } from 'react';

interface Props {
  placeholder: string;
  defaultValue: string;
}

interface Query {
  query: {
    value: string;
  };
}

export default function Search({ placeholder, defaultValue }: Props) {
  const router = useRouter();
  const inputRef = useRef<HTMLInputElement>(null);
  useDefaultValue(inputRef);

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const elements = event.currentTarget.elements as typeof event.currentTarget.elements & Query;
    const [owner, repo] = elements.query.value.split('/');
    router.push({
      pathname: '/',
      query: {
        owner: owner,
        repo: repo,
      },
    });
  };

  return (
    <form onSubmit={handleSubmit} className="flex flex-row my-5">
      <input
        ref={inputRef}
        required
        defaultValue={defaultValue}
        pattern="[a-zA-Z0-9-]+\/[a-zA-Z0-9-]+"
        name="query"
        className="p-2 flex-1"
        placeholder={placeholder}
        type="search"
      />
      <Button>Search</Button>
    </form>
  );
}
