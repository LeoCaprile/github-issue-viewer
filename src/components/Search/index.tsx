import Button from '@components/UI/Button';
import { useRouter } from 'next/router';
import { FormEvent, useEffect, useRef, useState } from 'react';

interface Props {
  placeholder: string;
}

interface Query {
  query: {
    value: string;
  };
}

export default function Search({ placeholder }: Props) {
  const router = useRouter();
  const [searchValue, setSearchValue] = useState<string>('facebook/react');
  const inputRef = useRef<HTMLInputElement>(null);
  const handleSearchValue = ({ target: { value } }: { target: { value: string } }) => setSearchValue(value);

  useEffect(() => {
    if (router.query.owner && router.query.repo) {
      setSearchValue(`${router.query.owner}/${router.query.repo}`);
    }
  }, [inputRef, router?.query.owner, router?.query.repo]);

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const elements = event.currentTarget.elements as typeof event.currentTarget.elements & Query;
    const [owner, repo] = elements.query.value.split('/');
    router.push({
      pathname: '/',
      query: {
        owner: owner,
        repo: repo,
      },
    });
  };

  return (
    <form onSubmit={handleSubmit} className="flex flex-row my-5">
      <label htmlFor="query"></label>
      <input
        value={searchValue}
        onChange={handleSearchValue}
        required
        pattern="[a-zA-Z0-9-]+\/[a-zA-Z0-9-]+"
        name="query"
        className="input input-bordered input-accent p-2 flex-1"
        placeholder={placeholder}
        type="search"
      />
      <Button>Search</Button>
    </form>
  );
}
