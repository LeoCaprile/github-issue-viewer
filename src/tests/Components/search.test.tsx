import { render, screen } from '@testing-library/react';
import Search from '../../components/Search';
import { useRouter } from '..';

describe('Search Component', () => {
  it('Renders normaly', () => {
    //Arrange
    useRouter.mockImplementationOnce(() => ({
      query: {},
    }));

    render(<Search placeholder="test" />);

    //Assert
    screen.getByDisplayValue('facebook/react');
  });

  it('Changes search value when query values change', async () => {
    useRouter.mockImplementation(() => ({
      route: '/',
      pathname: '/',
      query: { owner: 'coffe', repo: 'uwu' },
      asPath: '',
    }));
    const component = render(<Search placeholder="test" />);

    expect(component).toMatchSnapshot();
  });
});
