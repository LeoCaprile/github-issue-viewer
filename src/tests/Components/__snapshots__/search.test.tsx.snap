// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component Changes search value when query values change 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="flex flex-row my-5"
      >
        <label
          for="query"
        />
        <input
          class="p-2 flex-1"
          name="query"
          pattern="[a-zA-Z0-9-]+\\/[a-zA-Z0-9-]+"
          placeholder="test"
          required=""
          type="search"
          value="coffe/uwu"
        />
        <button
          class="border-white border-2 p-2 rounded-lg mx-3"
        >
          Search
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="flex flex-row my-5"
    >
      <label
        for="query"
      />
      <input
        class="p-2 flex-1"
        name="query"
        pattern="[a-zA-Z0-9-]+\\/[a-zA-Z0-9-]+"
        placeholder="test"
        required=""
        type="search"
        value="coffe/uwu"
      />
      <button
        class="border-white border-2 p-2 rounded-lg mx-3"
      >
        Search
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
